### PR TITLE
fr-BE and fr-CH support for ToWords and ToOrdinalWords PR

### LINKED ISSUE #538: Belgian french support for NumberToWords and NumberToOrdinalWords
In Walloon region we don't use same "soixante-dix" as French people but "septante" instead.

[See documentation here](http://www.languagesandnumbers.com/comment-compter-en-francais-belgique/fr/fra-bel/)

Already forked and implemented.
PR coming.

### LINKED ISSUE #541: Swiss french support for NumberToWords and NumberToOrdinalWords
Same as #538 for fr-CH

### LINKED ISSUE #548: french NumberToWords does not take grammatical gender into account
In french, the quantifier should agrees with the noun both in number and **in gender**. However, this rule only applies to number ending with 1's (un) with the exception of 11's (onze) such as 11, 71(60+11) and 91(80+11).

With the expression : 
```
new [] { 1, 21, 31, 41, 51, 61, 81 }
    .Select(x => x.ToWords(GrammaticalGender.Feminine, new CultureInfo("fr-Ca")))
```
What we get | What we expect
--- | ---
un | une 
vingt et un | vingt et une 
trente et un | trente et une 
quarante et un | quarante et une 
cinquante et un | cinquante et une 
soixante et un | soixante et une 
quatre-vingt-un | quatre-vingt-une

### LINKED ISSUE #557: NumberToWords in French is wrong for a few numbers
Hi,

Edit:  Added a few edge cases with thousands and millions...

In French, twenty (vingt) and one hundred (cent) have to be accorded to plural, but only when it's not followed by any other number.  Please note that in French, eighty is quatre-vingts (four times twenty...)...

Here are a few examples:

Number|Actual result|Expected value|Correct?
------|-------------|--------------|--------
80|quatre-vingt|quatre-vingts|False
81|quatre-vingt-un|quatre-vingt-un|True
400|quatre cent|quatre cents|False
401|quatre cent un|quatre cent un|True
480|quatre cent quatre-vingt|quatre cent quatre-vingts|False
80000|quatre-vingt mille|quatre-vingt mille|True
80000000|quatre-vingt millions|quatre-vingts millions|False
80080080|quatre-vingt millions quatre-vingt mille quatre-vingt|quatre-vingts millions quatre-vingt mille quatre-vingts|False
200200200|deux cent millions deux cent mille deux cent|deux cents millions deux cent mille deux cents|False
200200202|deux cent millions deux cent mille deux cent deux|deux cents millions deux cent mille deux cent deux|False

Thank you for your help.

### PR DESCRIPTION
**Pull request by jaxx-rep**
Add fr-BE and fr-CH eccentricities

Fixes #538 
Fixes #541  
Fixes #548 
Fixes #557

### CODE REVIEW

File: src/Humanizer/Localisation/NumberToWords/FrenchNumberToWordsConverter.cs [hunk 1]
```diff
@@ -111,5 +129,70 @@ public override string ConvertToOrdinal(int number, GrammaticalGender gender)
             convertedNumber += "ième";
             return convertedNumber;
         }
+
+        private string[] GetCultureTensMap()
```
Comment (hazzik) [inline src/Humanizer/Localisation/NumberToWords/FrenchNumberToWordsConverter.cs]: Can we make this virtual and implement in Culture-specific converters?

**Comment by jaxx-rep** [comment]
There is more than just this one method to change, if i remember correctly, there is at least one more method and a modification in 'Convert' too.

I'll take a look when possible.

**Comment by jaxx-rep** [comment]
@hazzik no virtual, but did you prefer this ?

I did it this way because converter code is almost identical, except for numbers < 100.

**Comment by ghigad** [comment]
Will this also work for other declinations of the fr locale, like fr-FR and fr-CA?

File: src/Humanizer.Tests.Shared/Localisation/fr/NumberToWordsTests.cs [hunk 2]
```diff
@@ -59,6 +59,20 @@ public void ToWords(int number, string expected)
...
+        [InlineData(121, "cent vingt et un", GrammaticalGender.Masculine)]
+        [InlineData(10121, "dix mille cent vingt et une", GrammaticalGender.Feminine)]
+        [InlineData(10121, "dix mille cent vingt et un", GrammaticalGender.Masculine)]
+        public void ToWordsWithGender(int number, string expected, GrammaticalGender gender)
+        {
+            Assert.Equal(expected, number.ToWords(gender));
+        }
+
+        [Theory]
         [InlineData(0, "zéroième")]
```
Comment (aloisdg) [inline src/Humanizer.Tests.Shared/Localisation/fr/NumberToWordsTests.cs]: I would prefer [zérotième](https://fr.wiktionary.org/wiki/z%C3%A9roti%C3%A8me).

**Comment by jaxx-rep** [comment]
Initially there was only 'fr' implemented with the rules for fr-FR.
I added support for language who don't follow the same rules set.

With fr-CA there is nothing more to do because it follow fr-FR rules (as far i can see)
It will defaults to fr implementation.

fr-BE -> FrenchBelgianNumberToWordsConverter
fr-CH -> FrenchSwissNumberToWordsConverter
fr / fr-* -> FrenchNumberToWordsConverter

TL;DR: Yep, it works, the fr implementation is the default fallback, which follow fr-FR rules.

----

Anyway, i just included a fix for #548 as well

File: src/Humanizer.Tests.Shared/Localisation/fr/NumberToWordsTests.cs (hunk 2 above)
Comment (jaxx-rep) [inline src/Humanizer.Tests.Shared/Localisation/fr/NumberToWordsTests.cs]: Damn xD

Thanks for pointing this one !

Comment (aloisdg) [inline src/Humanizer.Tests.Shared/Localisation/fr/NumberToWordsTests.cs]: [zéroième](https://fr.wiktionary.org/wiki/z%C3%A9roi%C3%A8me) is in fact fine, but rare. :)

Comment (jaxx-rep) [inline src/Humanizer.Tests.Shared/Localisation/fr/NumberToWordsTests.cs]: Done : f46d7d0

**Comment by aloisdg** [comment]
By the way, I propose that we continue to ignore "second/seconde" because it is a bit of a mess and "deuxième" is quite accepted all the time.

[wiki entry](https://fr.wikipedia.org/wiki/Adjectif_num%C3%A9ral#Morphologie_de_l.27ordinal) on this subject.

**Comment by jaxx-rep** [comment]
Thanks for the reference.
The rules for this cannot really be inferred without a grammatical context, which is _a little bit_ out of scope ;)

Stick to "deuxiéme" for now..

**Comment by aloisdg** [comment]
"deuxième" ;)

File: src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                parts.Add(string.Format("{0} million{1}",
+                   Convert(number/1000000),
+                   number/1000000 == 1 ? string.Empty : "s"));
+
+                number %= 1000000;
+            }
+
+            if ((number/1000) > 0)
+            {
+                parts.Add(string.Format("{0}mille",
```
Comment (hazzik) [inline src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs]: Should not this be `"{0} mille"`? Note the space.

File: src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+            if (number == 0)
+                convertedNumber += "t";
+
+            convertedNumber = convertedNumber.TrimEnd('e');
+            convertedNumber += "ième";
+            return convertedNumber;
+        }
+
+        protected abstract void LastPart(ref List<string> parts, ref int number);
```
Comment (hazzik) [inline src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs]: don't need `ref` in front of `List<string> parts`

File: src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+                number %= 1000;
+            }
+
+            if ((number/100) > 0)
+            {
+                if (number < 200)
+                    parts.Add("cent");
+                else
+                    parts.Add(string.Format("{0} cent{1}",
```
Comment (hazzik) [inline src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs]: There is no tests for this branch

**Comment by hazzik** [comment]
What shall be a result of 1000000, feminine?

File: src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs (hunk 3 above)
Comment (aloisdg) [inline src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs]: @hazzik I dont know. It should be, maybe. I refactored. I didn't change the original logic. Tests should check this ;)

Comment (hazzik) [inline src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs]: They fail

Comment (aloisdg) [inline src/Humanizer/Localisation/NumberToWords/FrenchAbstractNumberToWordsConverter.cs]: @hazzik Just checked. I missed place the whitespace. I wrote it after "mille".

**Comment by hazzik** [comment]
Replaced by #559
